aptos-move/framework/src/natives/cryptography/groups/arkworks_backend.rs

```diff
@@ -0,0 +1,1896 @@
+        self.fr_store.get(handle).unwrap()
+    }
+
+    pub fn add_g1_point(&mut self, p0: G1Projective) -> usize {
+        let ret = self.g1_point_store.len();
+        self.g1_point_store.push(p0);
+        ret
+    }
+
+    pub fn get_g1_point(&self, handle: usize) -> &G1Projective {
+        self.g1_point_store.get(handle).unwrap()
+    }
+
+    pub fn add_g2_point(&mut self, p0: G2Projective) -> usize {
+        let ret = self.g2_point_store.len();
+        self.g2_point_store.push(p0);
+        ret
+    }
+
+    pub fn get_g2_point(&self, handle: usize) -> &G2Projective {
+        self.g2_point_store.get(handle).unwrap()
+    }
+
+    pub fn add_gt_point(&mut self, point: Fq12) -> usize {
+        let ret = self.gt_point_store.len();
+        self.gt_point_store.push(point);
+        ret
+    }
+
+    pub fn get_gt_point(&self, handle: usize) -> &Fq12 {
+        self.gt_point_store.get(handle).unwrap()
+    }
+}
+
+fn serialize_element_uncompressed_internal(
+    gas_params: &GasParameters,
+    context: &mut NativeContext,
+    ty_args: Vec<Type>,
+    mut args: VecDeque<Value>,
+) -> PartialVMResult<NativeResult> {
+    assert_eq!(1, ty_args.len());
+    let type_tag = context
+        .type_to_type_tag(ty_args.get(0).unwrap())?
+        .to_string();
+    let handle = pop_arg!(args, u64) as usize;
+    match type_tag.as_str() {
+        "0x1::groups::BLS12_381_G1" => {
+            let mut buf = vec![];
+            context
+                .extensions()
+                .get::<Bls12381Context>()
+                .get_g1_point(handle)
+                .into_affine()
+                .serialize_uncompressed(&mut buf)
+                .unwrap();
+            Ok(NativeResult::ok(
+                gas_params.bls12_381.g1_proj_to_affine + gas_params.bls12_381.g1_affine_serialize_uncompressed,
+                smallvec![Value::vector_u8(buf)],
+            ))
+        }
+        "0x1::groups::BLS12_381_G2" => {
+            let mut buf = vec![];
+            context
+                .extensions()
+                .get::<Bls12381Context>()
+                .get_g2_point(handle)
+                .into_affine()
+                .serialize_uncompressed(&mut buf)
+                .unwrap();
+            Ok(NativeResult::ok(
+                gas_params.bls12_381.g2_proj_to_affine + gas_params.bls12_381.g2_affine_serialize_uncompressed,
+                smallvec![Value::vector_u8(buf)],
+            ))
+        }
+        "0x1::groups::BLS12_381_Gt" => {
+            let mut buf = vec![];
+            context
+                .extensions()
+                .get::<Bls12381Context>()
+                .get_gt_point(handle)
+                .serialize_uncompressed(&mut buf)
+                .unwrap();
+            Ok(NativeResult::ok(
+                gas_params.bls12_381.fq12_serialize,
+                smallvec![Value::vector_u8(buf)],
+            ))
+        }
+        _ => {
+            Ok(NativeResult::err(InternalGas::zero(), E_UNKNOWN_GROUP))
```

Comment:
So, abort codes must be wrapped by an error code type from `move-stdlib/sources/error.move`.

In your case, the three error codes you've defined all seem to pertain to an [`INVALID_ARGUMENT` type of error](https://github.com/aptos-labs/aptos-core/blob/3ed9bb266bf70b4e768139b3034c049bd318e24c/aptos-move/framework/move-stdlib/sources/error.move#L25).

So you should declare your codes like in [this example](https://github.com/aptos-labs/aptos-core/blob/7d7822a74c0947ac7c743c0cf19cbd5cac0a93bf/aptos-move/framework/src/natives/cryptography/secp256k1.rs#L20))

You will need to declare the codes in your `groups.move` file too, although there you will declare them without the error-type wrapping, just like you originally (mis)declared them in the Rust file.

This way, we can indicate somewhere in the comments that, for example:
```
/// All functions in this module abort with an `std::error::invalid_argument(E_UNKNOWN_GROUP)` abort code when the wrong group type argument is given.
```
Plus, whatever other comments we will need to add.

Lastly, you'll need a `E_UNKNOWN_FIELD` after changing the functions to use `Scalar<F>` rather than `Scalar<G>`. And you will even need a `E_MISMATCHED_GROUP_AND_FIELD` for functions whose group and field do not match (i.e., do not have the right types)

Reply:
No longer need to deal with this in rust backend...